molgenis-data-security/src/main/java/org/molgenis/data/security/auth/GroupService.java

```diff
@@ -86,6 +102,126 @@ public void persist(GroupValue groupValue)
+	 */
+	public void grantPermissions(GroupValue groupValue)
+	{
+		PackageIdentity packageIdentity = new PackageIdentity(groupValue.getRootPackage().getName());
+		GroupIdentity groupIdentity = new GroupIdentity(groupValue.getName());
+		groupValue.getRoles().forEach(roleValue ->
+		{
+			PermissionSet permissionSet = DEFAULT_ROLES.get(roleValue.getLabel());
+			Sid roleSid = createRoleSid(roleValue.getName());
+			permissionService.grant(packageIdentity, permissionSet, roleSid);
+			permissionService.grant(groupIdentity, permissionSet, roleSid);
+		});
+		if (groupValue.isPublic())
+		{
+			permissionService.grant(groupIdentity, READ, createAuthoritySid(AUTHORITY_USER));
+		}
+	}
+
+	@RunAsSystem
+	public Collection<Group> getGroups()
+	{
+		return dataService.findAll(GroupMetadata.GROUP, Group.class).collect(Collectors.toList());
+	}
+
+	/**
+	 * Get the group entity by its unique name.
+	 *
+	 * @param groupName unique group name
+	 * @return group with given name
+	 * @throws UnknownEntityException in case no group with given name can be retrieved
+	 */
+	@RunAsSystem
+	public Group getGroup(String groupName)
+	{
+		Group group = dataService.query(GroupMetadata.GROUP, Group.class).eq(GroupMetadata.NAME, groupName).findOne();
+		if (group == null)
+		{
+			throw new UnknownEntityException(groupMetadata, groupMetadata.getAttribute(GroupMetadata.NAME), groupName);
+		}
+		return group;
+	}
+
+	/**
+	 * Add member to group.
+	 * User can only be added to a role that belongs to the group.
+	 * The user can only have a single role within the group
+	 *
+	 * @param group group to add the user to in the given role
+	 * @param user user to be added in the given role to the given group
+	 * @param role role in which the given user is to be added to given group
+	 */
+	@RunAsSystem
+	public void addMember(final Group group, final User user, final Role role)
+	{
+		ArrayList<Role> groupRoles = newArrayList(group.getRoles());
+		Collection<RoleMembership> memberships = roleMembershipService.getMemberships(groupRoles);
+		boolean isGroupRole = groupRoles.stream().anyMatch(gr -> gr.getName().equals(role.getName()));
+
+		if(!isGroupRole)
+		{
+			throw new NotAValidGroupRoleException(role, group);
+		}
+
+		boolean isMember = memberships.stream().parallel().anyMatch(m -> m.getUser().equals(user));
+
+		if(isMember)
+		{
+			throw new IsAlreadyMemberException(user, group);
+		}
+
+		roleMembershipService.addUserToRole(user, role);
+	}
+
+	@RunAsSystem
+	public void removeMember(final Group group, final User user)
+	{
+		ArrayList<Role> groupRoles = newArrayList(group.getRoles());
+		final RoleMembership member = roleMembershipService.getMemberships(groupRoles)
+														   .stream()
+														   .filter(m -> m.getUser().getId().equals(user.getId()))
+														   .findFirst()
+														   .orElseThrow(() -> new UnknownEntityException(
+																   roleMembershipMetadata,
```

Comment:
i would argue this was clearer before, just because every thing can be chained does not mean is should.

Reply:
https://dzone.com/articles/java-8-optional-replace-your-get-calls is what I was thinking about

Reply:
yes, the article does not suggest to chain the everything, it closes with a comment about the needed functionality not being available until java 9, 

```Unfortunately this is yet to come It will be available in Java 9.

Until then we have to write something like:

if(optional.isPresent()){
    doSomething(optional.get());
} else {
    doSomethingElse();
}
There are cases in which you are allowed to use get() and isPresent() but use them with a grain of salt.```

Reply:
I've split off the creation of the exception and reused the existing findRoleMembership method.